staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,765 @@
+                    desc_request.len(),
+                ));
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioCanFrame>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            CanController::print_can_frame(request);
+
+            let msg_type = request.msg_type.to_native();
+            let mut can_id = request.can_id.to_native();
+            let mut flags = request.flags.to_native();
+            let mut length = request.length.to_native();
+
+            if msg_type != VIRTIO_CAN_TX {
+                warn!("TX: Message type 0x{:x} unknown\n", msg_type);
+                return Err(Error::UnexpectedCanMsgType(msg_type));
+            }
+
+            if (flags & VIRTIO_CAN_FLAGS_FD) != 0 {
+                if length > 64 {
+                    trace!("Cut sdu_len from {:?} to 64\n", request.length);
+                    length = 64;
+                }
+            } else if length > 8 {
+                trace!("Cut sdu_len from {:?} to 8\n", request.length);
+                length = 8;
+            }
+
+            /*
+             * Copy Virtio frame structure to qemu frame structure and
+             * check while doing this whether the frame type was negotiated
+             */
+            if (flags & VIRTIO_CAN_FLAGS_EXTENDED) != 0 {
+                flags &= CAN_EFF_MASK;
+                flags |= CAN_EFF_FLAG;
+            } else {
+                flags &= CAN_SFF_MASK;
+            }
+
+            if (flags & VIRTIO_CAN_FLAGS_RTR) != 0 {
+                if !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC)
+                    || !self.check_features(VIRTIO_CAN_F_RTR_FRAMES)
+                {
+                    warn!("TX: RTR frames not negotiated");
+                    return Err(Error::UnexpectedRtrFlag);
+                }
+                can_id |= flags | CAN_RTR_FLAG;
+            }
+
+            if (flags & VIRTIO_CAN_FLAGS_FD) != 0 {
+                if !self.check_features(VIRTIO_CAN_F_CAN_FD) {
+                    warn!("TX: FD frames not negotiated\n");
+                    return Err(Error::UnexpectedFdFlag);
+                }
+                flags |= CAN_FRMF_TYPE_FD;
+            } else {
+                if !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC) {
+                    warn!("TX: Classic frames not negotiated\n");
+                    return Err(Error::UnexpectedClassicFlag);
+                }
+                flags = 0;
+            }
+
+            let corrected_request = VirtioCanFrame {
+                msg_type: msg_type.into(),
+                can_id: can_id.into(),
+                length: length.into(),
+                reserved: 0.into(),
+                flags: flags.into(),
+                sdu: request.sdu[0..64].try_into().unwrap(),
+            };
+
+            let desc_response = descriptors[1];
+            if !desc_response.is_write_only() {
```

Comment:
Maybe it is better to check this early, for example, when checking for `descriptors[0]`?

Reply:
Yes indeed, I just added a simple test above that point, to make sure `descriptor[1]` exists before accessing it.
```test
if descriptors.len() != 2 {
    return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
}
```

Reply:
Cool, I think we can get rid of all that code that checks descriptors.len() and the flags by using Reader/Writer. I can help with that if required.

Reply:
I could try to do it, but let me understand a bit more what this means.
So the idea is to have a function `fn descriptor_read_write_checks (descriptos, ...)` which returns `Ok((descriptor[0], descriptor[2])`, if the length is the required and all read/write tests are successful. Is that correct?

Reply:
A new version implementing the idea of reader/writer based on the code you provided have been pushed.
To do so I needed to include your version of vhost, vhost-user-backend into the Cargo.toml.

I believe including this update we have addressed all received comments till now.
Would you think that we are close to possibly accept this code into vhost-device repo?

Also an open point from my side is that the code is failing into the pipeline's coverage test.
As far as I see that is happening because the unit tests cover ~1% more than the
predefined upper limit:
```
Current code coverage (63.16%) deviates by 1.01% from the previous code coverage 62.15%
```
Is that a problem, or we can just increase the upper limit into the `coverage_config_x86_64.json`?

Reply:
> To do so I needed to include your version of vhost, vhost-user-backend into the Cargo.toml.

FYI I'm preparing a new release of virtio-queue: https://github.com/rust-vmm/vm-virtio/pull/292

> Would you think that we are close to possibly accept this code into vhost-device repo?

I think so :-) thanks for addressing the reviews. Let's give other a bit more time to review, then we can merge it (we need to wait the virtio-queue release anyway)

> Is that a problem, or we can just increase the upper limit into the coverage_config_x86_64.json?

Increasing the value `coverage_config_x86_64.json` is always a great sign that you implemented tests, so is not a problem at all increasing it in this PR.

Reply:
Thanks for the feedback, I will wait so for the virtio-queue update

Reply:
I think you can already update virtio-queue, however, we still need to wait for a release of vhost and vhost-user-device. For these creates, you still need to point to the git repository.

Reply:
Great! I have updated the virtio-queue crate's version in the latest push

Reply:
Hello @MatiasVara, I just wanted to let you know that the code works well with the latest version of `vhost: 0.11` and `vhost-user-backend: 0.15` crates, so I have removed the git repositories from the Cargo.toml.